From 63af0d38da8eefb208c0484da52dc914c335c780 Mon Sep 17 00:00:00 2001
From: Michael Wachenschwanz <mwachens@google.com>
Date: Wed, 15 May 2019 22:58:15 -0700
Subject: [PATCH 254/306] Clear the Parcel before writing an exception during a
 transaction

This prevents any object data from being accidentally overwritten by the
exception, which could cause unexpected malformed objects to be sent
across the transaction.

Test: atest CtsOsTestCases:ParcelTest#testExceptionOverwritesObject
Bug: 34175893
Change-Id: Iaf80a0ad711762992b8ae60f76d861c97a403013
Merged-In: Iaf80a0ad711762992b8ae60f76d861c97a403013
(cherry picked from commit f8ef5bcf21c87d8617f5e11810cc94350298d114)
---
 core/java/android/os/Binder.java | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/core/java/android/os/Binder.java b/core/java/android/os/Binder.java
index 8e8565c3574..4db8d1ea64d 100644
--- a/core/java/android/os/Binder.java
+++ b/core/java/android/os/Binder.java
@@ -740,6 +740,8 @@ public class Binder implements IBinder {
                     Log.w(TAG, "Caught a RuntimeException from the binder stub implementation.", e);
                 }
             } else {
+                // Clear the parcel before writing the exception
+                reply.setDataSize(0);
                 reply.setDataPosition(0);
                 reply.writeException(e);
             }
-- 
2.17.1

